### PR TITLE
Gradle plugin extension (config)

### PR DESCRIPTION
**Pull request by AlexanderLitus**

### CODE REVIEW

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy [hunk 1]
```diff
@@ -1,22 +1,156 @@
-package org.spine3.gradle;
+package org.spine3.gradle
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy]: So, we don't use semicolons anymore?

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy [hunk 2]
```diff
@@ -1,22 +1,156 @@
...
-import org.spine3.gradle.lookup.proto.ProtoToJavaMapperPlugin;
+import com.google.common.collect.ImmutableList
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.spine3.gradle.cleaning.CleaningPlugin
+import org.spine3.gradle.failures.FailuresGenPlugin
+import org.spine3.gradle.lookup.enrichments.EnrichmentLookupPlugin
+import org.spine3.gradle.lookup.proto.ProtoToJavaMapperPlugin
+
+import static java.util.Collections.*
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy]: Please expand asterisk import

Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy]: As we use them in the code below, I think that code formatting removed semicolons here

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy [hunk 3]
```diff
@@ -1,22 +1,156 @@
...
-        new FailuresGenPlugin().apply(target);
+    void apply(Project project) {
+        project.extensions.create("spineProtobuf", Extension.class);
+        new CleaningPlugin().apply(project);
+        new ProtoToJavaMapperPlugin().apply(project);
+        new EnrichmentLookupPlugin().apply(project);
+        new FailuresGenPlugin().apply(project);
+    }
+
+    public static String getMainTargetGenResourcesDir(Project project) {
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy]: It seems a bit strange to have public static stuff in plugins. I'd prefer having a utility class with these

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy (hunk 2 above)
Comment (AlexanderLitus) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/ProtobufPlugin.groovy]: Unfortunately, IDEA does not support such feature for Groovy... And doing this manually seems like  pain-in-ass)

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/cleaning/CleaningPlugin.groovy [hunk 4]
```diff
@@ -4,59 +4,31 @@ import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.api.Task
 
+import static org.spine3.gradle.ProtobufPlugin.getDirsToClean
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/cleaning/CleaningPlugin.groovy]: Semicolons

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/cleaning/CleaningPlugin.groovy [hunk 5]
```diff
@@ -4,59 +4,31 @@ import org.gradle.api.Plugin
...
-    private String projectPath;
-
-    private def defaultDirsToClean = ["generated"];
-
     /**
-     * Applied to project.
-     *
-     * <p>Adds :preClean task.
-     * <p>Is executed before :clean task.
+     * Adds `:preClean` task, whic is executed before `:clean` task.
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/cleaning/CleaningPlugin.groovy]: "which"

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/failures/FailuresGenPlugin.groovy [hunk 6]
```diff
@@ -1,13 +1,13 @@
...
-
-import com.google.protobuf.DescriptorProtos;
-import groovy.util.logging.Slf4j;
-import org.gradle.api.Plugin;
-import org.gradle.api.Project;
-import org.gradle.api.Task;
-
-import static com.google.protobuf.DescriptorProtos.FileDescriptorProto;
-
+package org.spine3.gradle.failures
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/failures/FailuresGenPlugin.groovy]: Semicolons

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/failures/FailuresGenPlugin.groovy [hunk 7]
```diff
@@ -57,38 +59,25 @@ class FailuresGenPlugin implements Plugin<Project> {
...
-                                                             boolean noDescriptorsWarning) {
-        final String descFilePath = "${projectPath}/build/descriptors/${descSourceSet}.desc";
-        final List<FileDescriptorProto> failureDescriptors = new ArrayList<>();
-
-        if (!new File(descFilePath).exists()) {
-            if (!noDescriptorsWarning) {
-                log.warn("Please enable descriptor set generation. See an appropriate section at https://github.com/google/protobuf-gradle-plugin/blob/master/README.md#customize-code-generation-tasks");
+    private List<FileDescriptorProto> getFailureProtoFileDescriptors(String descFilePath) {
+        final List<FileDescriptorProto> failureDescriptors = new LinkedList<>();
+        def allDescriptors = getProtoFileDescriptors(descFilePath);
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/failures/FailuresGenPlugin.groovy]: final?

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy [hunk 8]
```diff
@@ -18,16 +18,16 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-package org.spine3.gradle.lookup.enrichments;
+package org.spine3.gradle.lookup.enrichments
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy]: Semicolons

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy [hunk 9]
```diff
@@ -18,16 +18,16 @@
...
-import groovy.util.logging.Slf4j;
-import org.gradle.api.Plugin;
-import org.gradle.api.Project;
-import org.gradle.api.Task;
-import org.spine3.gradle.lookup.PropertiesWriter;
-
-import static com.google.protobuf.DescriptorProtos.FileDescriptorProto;
-import static com.google.protobuf.DescriptorProtos.FileDescriptorSet;
+import static com.google.protobuf.DescriptorProtos.FileDescriptorProto
+import static org.spine3.gradle.ProtobufPlugin.*
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy]: Please expand

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/proto/ProtoToJavaMapperPlugin.groovy [hunk 10]
```diff
@@ -25,10 +25,15 @@ import groovy.util.logging.Slf4j;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
-import org.spine3.gradle.lookup.PropertiesWriter;
+import org.spine3.gradle.util.PropertiesWriter;
 
 import java.util.regex.Matcher;
-import java.util.regex.Pattern;
+import java.util.regex.Pattern
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/proto/ProtoToJavaMapperPlugin.groovy]: Semicolon

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy (hunk 8 above)
Comment (AlexanderLitus) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy]: I'd better kill them all)

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/util/DescriptorSetUtil.groovy [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.gradle.util;
+
+import com.google.common.base.Predicate
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/util/DescriptorSetUtil.groovy]: Semicolon

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy (hunk 8 above)
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy]: So, we're not gonna use semicolons in groovy imports, right?

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy [hunk 12]
```diff
@@ -11,8 +11,8 @@
  * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
  * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
  * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
- * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
- * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES LOSS OF USE,
+ * DATA, OR PROFITS OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy]: Is this semicolon removal confirmed?

Comment (AlexanderLitus) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy]: Updated all licenses.

**Comment by AlexanderLitus** [comment]
Hmm, the code looks really neat without semicolons.

File: protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy [hunk 13]
```diff
@@ -54,14 +53,14 @@ public class EnrichmentLookupPlugin implements Plugin<Project> {
...
-            findEnrichmentsAndWriteProps(ExtensionUtil.getMainTargetGenResourcesDir(project), ExtensionUtil.getMainDescriptorSetPath(project))
+            findEnrichmentsAndWriteProps(org.spine3.gradle.Extension.getMainTargetGenResourcesDir(project), org.spine3.gradle.Extension.getMainDescriptorSetPath(project))
         }
         findEnrichmentsTask.dependsOn("compileJava")
         final Task processResources = project.getTasks().getByPath("processResources")
         processResources.dependsOn(findEnrichmentsTask)
 
         final Task findTestEnrichmentsTask = project.task("findTestEnrichments") << {
-            findEnrichmentsAndWriteProps(ExtensionUtil.getTestTargetGenResourcesDir(project), ExtensionUtil.getTestDescriptorSetPath(project))
+            findEnrichmentsAndWriteProps(org.spine3.gradle.Extension.getTestTargetGenResourcesDir(project), org.spine3.gradle.Extension.getTestDescriptorSetPath(project))
```
Comment (mikhailmikhaylov) [inline protobuf-plugin/plugin/src/main/groovy/org/spine3/gradle/lookup/enrichments/EnrichmentLookupPlugin.groovy]: Do we use FQN because of collisions?

**Comment by mikhailmikhaylov** [comment]
LGTM
@alexander-yevsyukov Please confirm

**Comment by alexander-yevsyukov** [comment]
@AlexanderLitus, please create a Wiki page with code style for Gradle and Groovy in general. Have it in core-java for now. We can move it into a separate repository later, if needed. Please do it outside of the scope of this PR. This PR LGTM.

**Pull request merged**